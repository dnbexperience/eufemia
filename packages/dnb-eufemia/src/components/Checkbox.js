/**
 * ATTENTION: This file is auto generated by using "prepareTemplates".
 * Do not change the content!
 *
 */

/**
 * Library Index checkbox to autogenerate all the components and patterns
 * Used by "prepareCheckboxs"
 */

import Checkbox from './checkbox/Checkbox'
export * from './checkbox/Checkbox'
export default Checkbox
