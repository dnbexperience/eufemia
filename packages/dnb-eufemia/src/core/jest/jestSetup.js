/**
 * Jest Setup
 *
 */

import { axe, toHaveNoViolations } from 'jest-axe'
import fakeProps from 'react-fake-props'
import { shallow, mount, render } from './enzyme'
import ReactDOMServer from 'react-dom/server'
import fs from 'fs-extra'
import onceImporter from 'node-sass-once-importer'
import path from 'path'
import sass from 'node-sass'
import { toBeType } from 'jest-tobetype'
import toJson from 'enzyme-to-json'

export {
  fakeProps, // we have also our own replacement function called "fakeAllProps"
  shallow,
  mount,
  render,
  toJson,
  axe,
  toHaveNoViolations
}

global.IS_TEST = true
if (typeof window !== 'undefined') {
  window.IS_TEST = true
}

expect.extend({ toBeType })
expect.extend(toHaveNoViolations)

export const loadScss = (file, options = {}) => {
  try {
    const sassResult = sass.renderSync({
      file,
      includePaths: [path.resolve(__dirname, '../../style/core/')],
      importer: [onceImporter()],
      ...options
    })
    return String(sassResult.css)
  } catch (e) {
    console.error('loadScss error:', e)
    return e
  }
}

export const mockGetSelection = () => {
  let memory
  Object.defineProperty(window.navigator, 'clipboard', {
    configurable: true,
    value: {
      writeText: jest.fn().mockImplementation((v) => {
        memory = v
        return Promise.resolve(v)
      }),
      readText: jest.fn().mockImplementation(() => Promise.resolve(memory))
    }
  })

  const mockRange = new (class Range {
    constructor() {
      this.startContainer = {
        parentNode: document.createElement('div')
      }
    }
    getElement() {
      return this.node
    }
    insertNode(elem) {
      this.node = document.createElement('div')
      this.node.appendChild(elem)
      return this
    }
    cloneRange() {
      return this
    }
  })()

  let ranges = [mockRange]

  const mockValue = '1234.56'
  let value = mockValue
  let rangeCount = 9

  class RangeObj {
    rangeCount = rangeCount
    toString = () => value
    addRange(range = mockRange) {
      value = mockValue
      ranges.push(range)
      rangeCount = ranges.length
    }
    getRangeAt(index) {
      return ranges[index]
    }
    removeAllRanges() {
      value = ''
      ranges = []
      rangeCount = ranges.length
    }
  }

  Object.defineProperty(window, 'getSelection', {
    configurable: true,
    value: () => {
      return new RangeObj()
    }
  })
}

export const loadImage = async (imagePath) =>
  await fs.readFile(path.resolve(imagePath))

export const toHtml = (Component) =>
  ReactDOMServer.renderToStaticMarkup(Component)

export const axeComponent = async (...components) => {
  const html = components
    .filter((Component) =>
      // enzyme names the mounted wrapper: ReactWrapper
      /react/i.test(String(Component.constructor))
    )
    .map((Component) => toHtml(Component))
    .join('\n')

  return await axe(
    `<main>${html}</main>`,
    typeof components[1] === 'object' ? components[1] : null
  )
}

export function attachToBody() {
  let container = document.createElement('div')
  document.body.append(container)
  return container
}

function jestSetup() {}

export default jestSetup
