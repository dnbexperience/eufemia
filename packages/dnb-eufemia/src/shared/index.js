/**
 * Shared
 *
 */

import Context from './Context'
import Provider from './Provider'
import MediaQuery from './MediaQuery'
import useMediaQuery from './useMediaQuery'
import useMedia from './useMedia'

export { Context, Provider, MediaQuery, useMediaQuery, useMedia }

// deprecated can be removed in v10
export * from '../components/space/SpacingHelper'
