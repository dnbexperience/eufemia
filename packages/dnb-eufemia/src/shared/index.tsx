/**
 * Shared
 *
 */

export { default as Context } from './Context'
export { default as Provider } from './Provider'
export { default as Theme } from './Theme'
export { default as useTheme } from './useTheme'
export { default as VisibilityByTheme } from './VisibilityByTheme'
export { default as MediaQuery } from './MediaQuery'
export { default as useMediaQuery } from './useMediaQuery'
export { default as useMedia } from './useMedia'
export { default as useTranslation } from './useTranslation'
export { default as Translation } from './Translation'
export * from './Translation'

export type { ThemeNames } from './Theme'
