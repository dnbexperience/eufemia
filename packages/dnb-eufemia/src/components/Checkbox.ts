/**
 * ATTENTION: This file is auto generated by using "prepareTemplates".
 * Do not change the content!
 *
 */

/**
 * Library Index checkbox to autogenerate all the components and extensions
 * Used by "prepareCheckboxs"
 */

import Checkbox from './checkbox/Checkbox'
export * from './checkbox/Checkbox'
export default Checkbox
