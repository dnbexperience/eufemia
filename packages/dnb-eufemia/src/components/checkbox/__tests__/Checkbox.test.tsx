/**
 * Checkbox Test
 *
 */

import {
  render,
  screen,
  cleanup,
  fireEvent,
  waitFor,
} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'
import { axeComponent, loadScss } from '../../../core/jest/jestSetup'
import Checkbox, { CheckboxProps } from '../Checkbox'
import { Provider } from '../../../shared'

const props: CheckboxProps = {
  label: 'checkbox',
}

describe('Checkbox component', () => {
  it('has correct state after "change" trigger', () => {
    const { rerender } = render(<Checkbox {...props} />)

    // default checked value has to be false
    expect(
      (screen.getByRole('checkbox') as HTMLInputElement).checked
    ).toBe(false)

    screen.getByRole('checkbox').click()
    expect(
      (screen.getByRole('checkbox') as HTMLInputElement).checked
    ).toBe(true)

    screen.getByRole('checkbox').click()
    expect(
      (screen.getByRole('checkbox') as HTMLInputElement).checked
    ).toBe(false)

    // also check if getDerivedStateFromProps sets the state as expected
    rerender(<Checkbox {...props} checked={true} />)

    expect(
      (screen.getByRole('checkbox') as HTMLInputElement).checked
    ).toBe(true)

    const value = 'new value'
    rerender(<Checkbox {...props} checked={true} value={value} />)
    expect((screen.getByRole('checkbox') as HTMLInputElement).value).toBe(
      value
    )
  })

  it('has "on_change" event which will trigger on a input change', () => {
    const myEvent = jest.fn()
    render(<Checkbox onChange={myEvent} checked={false} />)
    screen.getByRole('checkbox').click()

    expect(myEvent.mock.calls.length).toBe(1)
    expect(myEvent.mock.calls[0][0]).toHaveProperty('checked')
    expect(myEvent.mock.calls[0][0].checked).toBe(true)
  })

  describe('controlled vs uncontrolled', () => {
    const ControlledVsUncontrolled = () => {
      const [checked, setChecked] = React.useState(true)
      const [random, setRandom] = React.useState(null)

      return (
        <>
          <Checkbox
            checked={checked}
            onChange={({ checked }) => setChecked(checked)}
          />
          <button id="set-state" onClick={() => setChecked(true)} />
          <button
            id="reset-undefined"
            onClick={() => setChecked(undefined)}
          />
          <button id="reset-null" onClick={() => setChecked(null)} />
          <button id="rerender" onClick={() => setRandom(Math.random())} />
          <code>{JSON.stringify({ checked, random })}</code>
        </>
      )
    }

    it('handles re-render + default state', () => {
      render(<ControlledVsUncontrolled />)

      fireEvent.click(document.querySelector('button#set-state'))
      expect(document.querySelector('input').checked).toBe(true)
      cleanup()
    })

    it('changes to false', () => {
      render(<ControlledVsUncontrolled />)

      fireEvent.click(document.querySelector('input'))
      expect(document.querySelector('input').checked).toBe(false)
      cleanup()
    })

    it('handles set it to true', () => {
      render(<ControlledVsUncontrolled />)

      fireEvent.click(document.querySelector('button#set-state'))
      expect(document.querySelector('input').checked).toBe(true)
      cleanup()
    })
    it('handles reset it with undefined to false', () => {
      render(<ControlledVsUncontrolled />)

      fireEvent.click(document.querySelector('button#reset-undefined'))

      expect(document.querySelector('input').checked).toBe(false)

      cleanup()
    })
    it('handles set to true + reset it with null to false', () => {
      render(<ControlledVsUncontrolled />)

      fireEvent.click(document.querySelector('button#set-state'))
      fireEvent.click(document.querySelector('button#reset-null'))

      expect(document.querySelector('input').checked).toBe(false)

      cleanup()
    })

    it('handles re-render + still false', async () => {
      render(<ControlledVsUncontrolled />)

      userEvent.click(document.querySelector('button#rerender'))

      waitFor(() => {
        expect(document.querySelector('input').checked).toBe(false)
      })

      cleanup()
    })
  })

  it('has a disabled attribute, once we set disabled to true', () => {
    render(<Checkbox disabled={true} />)

    expect(
      (screen.getByRole('checkbox') as HTMLInputElement).disabled
    ).toBe(true)
  })

  it('should support spacing props', () => {
    render(<Checkbox top="2rem" />)

    const element = document.querySelector('.dnb-checkbox')

    expect(Array.from(element.classList)).toEqual([
      'dnb-checkbox',
      'dnb-form-component',
      'dnb-space__top--large',
    ])
  })

  it('should inherit formElement vertical label', () => {
    render(
      <Provider
        formElement={{ label_direction: 'vertical', disabled: true }}
      >
        <Checkbox label="Label" />
      </Provider>
    )

    const element = document.querySelector('.dnb-checkbox')
    const attributes = Array.from(element.attributes).map(
      (attr) => attr.name
    )
    const inputElement = document.querySelector('.dnb-checkbox input')
    const inputAttributes = Array.from(inputElement.attributes).map(
      (attr) => attr.name
    )

    expect(attributes).toEqual(['class'])
    expect(inputAttributes).toEqual([
      'id',
      'name',
      'type',
      'class',
      'disabled',
      'aria-disabled',
      'value',
    ])
    expect(Array.from(element.classList)).toEqual([
      'dnb-checkbox',
      'dnb-form-component',
      'dnb-checkbox--label-position-right',
    ])
    expect(Array.from(inputElement.classList)).toEqual([
      'dnb-checkbox__input',
    ])
  })

  it('should validate with ARIA rules', async () => {
    const Comp = render(<Checkbox {...props} />)
    expect(await axeComponent(Comp)).toHaveNoViolations()
  })

  it('gets valid ref element', () => {
    let ref: React.RefObject<HTMLInputElement>

    function MockComponent() {
      ref = React.useRef()
      return <Checkbox id="unique" innerRef={ref} />
    }

    render(<MockComponent />)

    expect(ref.current.getAttribute('id')).toBe('unique')
    expect(ref.current.classList).toContain('dnb-checkbox__input')
    expect(ref.current.tagName).toBe('INPUT')
  })

  it('gets valid element when ref is function', () => {
    const ref: React.MutableRefObject<HTMLInputElement> = React.createRef()

    const refFn = (elem: HTMLInputElement) => {
      ref.current = elem
    }

    render(<Checkbox id="unique" innerRef={refFn} />)

    expect(ref.current.getAttribute('id')).toBe('unique')
    expect(ref.current.classList).toContain('dnb-checkbox__input')
    expect(ref.current.tagName).toBe('INPUT')
  })

  it('should use span element if defined', () => {
    render(<Checkbox element="span" />)
    expect(document.querySelector('.dnb-checkbox__input').tagName).toBe(
      'SPAN'
    )
    expect(
      document.querySelector('.dnb-checkbox__input').getAttribute('type')
    ).toBe('checkbox')
  })
})

describe('Checkbox scss', () => {
  it('has to match style dependencies css', () => {
    const css = loadScss(require.resolve('../style/deps.scss'))
    expect(css).toMatchSnapshot()
  })

  it('have to match default theme snapshot', () => {
    const css = loadScss(
      require.resolve('../style/themes/dnb-checkbox-theme-ui.scss')
    )
    expect(css).toMatchSnapshot()
  })
})
