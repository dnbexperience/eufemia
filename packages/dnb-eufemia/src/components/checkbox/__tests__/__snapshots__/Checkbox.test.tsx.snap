// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox scss has to match style dependencies css 1`] = `
"/*
* Used for snapshot testing
*
*/
/*
 * Utilities
 */
/*
* Used for snapshot testing
*
*/
/*
 * Utilities
 */
/*
 * FormLabel component
 *
 */
.dnb-form-label {
  display: inline-block;
  padding: 0;
  text-align: left;
  vertical-align: baseline;
  white-space: pre-wrap;
  word-break: normal;
  font-size: var(--font-size-basis);
  color: inherit;
}
.dnb-form-label[for] {
  width: fit-content;
}
.dnb-form-label--vertical {
  display: block;
  margin-right: 0;
  margin-bottom: 0.5rem;
}
.dnb-form-label--interactive {
  cursor: pointer;
}
.dnb-form-label[disabled] {
  cursor: not-allowed;
}

legend.dnb-form-label {
  display: inline-block;
}

/*
* Used for snapshot testing
*
*/
/*
 * Utilities
 */
/*
 * FormStatus component
 *
 */
/*
 * Utilities
 */
.dnb-form-status {
  --form-status-radius: 0.25rem;
  display: flex;
  opacity: 1;
}
.dnb-form-status__shell {
  display: flex;
  justify-content: flex-start;
  align-items: flex-start;
  min-width: inherit;
  border-radius: var(--form-status-radius);
}
.dnb-form-status__text {
  padding: 0.625rem 1rem;
  cursor: text;
  color: inherit;
  line-height: var(--line-height-small);
  font-size: var(--font-size-small);
  white-space: normal;
}
button .dnb-form-status__text {
  cursor: inherit;
}
.dnb-form-status__text .dnb-ul,
.dnb-form-status__text .dnb-anchor {
  font-size: inherit;
}
.dnb-form-status__text .dnb-ul:not([class*=dnb-space__left]) {
  padding-left: 1rem;
}
.dnb-spacing .dnb-form-status__text .dnb-ul:not([class*=dnb-space__bottom]) {
  margin-bottom: 0.5rem;
}
.dnb-spacing .dnb-form-status__text .dnb-ul:not([class*=dnb-space__bottom]) .dnb-li {
  margin-top: 0.5rem;
  margin-bottom: 0.5rem;
}
.dnb-icon + .dnb-form-status__text {
  padding-left: 0.5rem;
}
.dnb-form-status__shell > .dnb-icon {
  display: flex;
  justify-content: center;
  align-items: center;
  margin: 0.3333333em 0.3333333em 0.3333333em 0.6666666em;
}
.dnb-form-status__size--large .dnb-form-status__text {
  padding-top: 1.125rem;
  padding-bottom: 1.125rem;
}
.dnb-form-status__size--large .dnb-form-status__shell > .dnb-icon {
  margin-top: 0.6666666em;
  margin-bottom: 0.6666666em;
}
.dnb-form-status--stretch {
  flex-grow: 1;
}
.dnb-form-status--stretch .dnb-form-status__shell {
  width: 100%;
}
.dnb-form-status--stretch .dnb-form-status__text {
  max-width: 47rem;
}
.dnb-form-status[hidden] {
  display: none;
}

/*
* Checkbox component
*
*/
/*
 * Utilities
 */
:root {
  --checkbox-width--medium: 1.5rem;
  --checkbox-height--medium: 1.5rem;
}

.dnb-checkbox {
  --checkbox-border-radius: 0.25rem;
  --checkbox-gfx-border-radius__indeterminate: 0.125rem;
  --checkbox-width--large: 2rem;
  --checkbox-height--large: 2rem;
  --checkbox-gfx-height__indeterminate: 0.625rem;
  --checkbox-gfx-width__indeterminate: 0.625rem;
  --checkbox-gfx-height__indeterminate--large: 0.875rem;
  --checkbox-gfx-width__indeterminate--large: 0.875rem;
  --checkbox-border-width: 0.125rem;
  --checkbox-border-width--hover: 0.125rem;
  --checkbox-color-gfx-on: black;
  --checkbox-color-gfx-off: white;
  --checkbox-color-background-on: white;
  --checkbox-color-background-off: white;
  --checkbox-color-border-on: black;
  --checkbox-color-border-off: black;
  --checkbox-color-gfx--disabled: grey;
  --checkbox-color-background-on--disabled: lightgrey;
  --checkbox-color-background-off--disabled: lightgrey;
  --checkbox-color-border-on--disabled: grey;
  --checkbox-color-border-off--disabled: grey;
  --checkbox-color-gfx__indeterminate--disabled: grey;
  --checkbox-color-background--active: lightgrey;
  --checkbox-color-border--active: transparent;
  --checkbox-color-gfx--hover: grey;
  --checkbox-color-background--hover: white;
  --checkbox-color-border-on--hover: grey;
  --checkbox-color-border-off--hover: grey;
  --checkbox-color-gfx--focus: grey;
  --checkbox-color-background--focus: lightgrey;
  --checkbox-color-gfx--error: lavenderblush;
  --checkbox-color-gfx--error-contrast: red;
  --checkbox-color-background-on--error: red;
  --checkbox-color-background--error-contrast: lavenderblush;
  --checkbox-color-border--error: red;
  --checkbox-color-gfx-indeterminate--error: red;
  --checkbox-color-gfx--error--hover: red;
  --checkbox-color-background--error--hover: lavenderblush;
  --checkbox-color-border--error--hover: red;
  --checkbox-color-background-indeterminate: grey;
  --checkbox-color-gfx-indeterminate: darkgray;
  --checkbox-color-background-indeterminate--active: grey;
  --checkbox-color-border-indeterminate--active: white;
  --checkbox-color-gfx-indeterminate--active: white;
  --checkbox-bounding--medium: 1.75, 1.75;
  display: inline-flex;
  flex-direction: column;
  font-size: var(--font-size-small);
  line-height: var(--line-height-basis);
  /*
   * Color scheme
   */
  /** Normal state **/
  /** Disabled state **/
  /** Active state **/
  /** Hover state **/
  /** Focus state **/
  /** Error state **/
}
.dnb-checkbox__inner {
  display: inline-flex;
  flex-direction: column;
  align-self: center;
}
.dnb-checkbox__shell {
  user-select: none;
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  width: var(--checkbox-width--medium);
  height: var(--checkbox-height--medium);
}
.dnb-checkbox--large .dnb-checkbox__shell {
  width: var(--checkbox-width--large);
  height: var(--checkbox-height--large);
}
.dnb-checkbox__button {
  display: inline-block;
  border: var(--checkbox-border-width) solid transparent;
}
.dnb-checkbox__focus {
  display: none;
  outline: none;
}
html[data-whatinput=keyboard] .dnb-checkbox__focus {
  --border-color: var(--focus-ring-color);
  --border-width: var(--focus-ring-width);
  box-shadow: 0 0 0 var(--border-width) var(--border-color);
  border-color: transparent;
}
.dnb-checkbox__focus, .dnb-checkbox__button {
  position: relative;
  z-index: 4;
  width: calc(var(--checkbox-width--medium) - 0.25rem);
  height: calc(var(--checkbox-height--medium) - 0.25rem);
  border-radius: var(--checkbox-border-radius);
}
.dnb-checkbox--large .dnb-checkbox__focus, .dnb-checkbox--large .dnb-checkbox__button {
  width: var(--checkbox-width--large);
  height: var(--checkbox-height--large);
  border-radius: var(--checkbox-border-radius);
}
.dnb-checkbox__gfx {
  position: absolute;
  z-index: 5;
  top: auto;
  left: auto;
  width: calc(var(--checkbox-width--medium) - 0.5rem);
  height: calc(var(--checkbox-height--medium) - 0.5rem);
  shape-rendering: geometricprecision;
  transition: opacity 200ms ease-out, transform 200ms ease-out;
}
.dnb-checkbox--large {
  line-height: var(--checkbox-height--large);
}
.dnb-checkbox--large .dnb-checkbox__gfx {
  width: calc(var(--checkbox-width--large) - 0.5rem);
  height: calc(var(--checkbox-height--large) - 0.5rem);
}
.dnb-checkbox__input {
  opacity: 0;
  position: absolute;
  top: auto;
  left: auto;
  z-index: 6;
  width: var(--checkbox-width--medium);
  height: var(--checkbox-height--medium);
  margin: 0;
  padding: 0;
  border: 0;
  transform: scale(var(--checkbox-bounding--medium));
}
.dnb-checkbox--large .dnb-checkbox__input {
  width: var(--checkbox-width--large);
  height: var(--checkbox-height--large);
  transform: scale(1);
}
.dnb-checkbox__input:not([disabled]) {
  cursor: pointer;
}
.dnb-checkbox .dnb-form-label {
  margin-bottom: 0;
  margin-right: 0;
  margin-left: 0;
}
.dnb-checkbox__order {
  display: inline-flex;
  align-items: baseline;
}
.dnb-checkbox__suffix {
  order: 4;
}
.dnb-checkbox--label-position-left .dnb-checkbox__order .dnb-checkbox__inner {
  order: 2;
}
.dnb-checkbox--label-position-left .dnb-checkbox__order .dnb-form-label {
  order: 1;
  padding-right: 0.5rem;
}
.dnb-checkbox--label-position-left .dnb-checkbox__order .dnb-form-status {
  order: 3;
  margin-top: 0.5rem;
}
.dnb-checkbox--label-position-right .dnb-checkbox__order .dnb-checkbox__inner {
  order: 1;
}
.dnb-checkbox--label-position-right .dnb-checkbox__order .dnb-form-label {
  order: 2;
  padding-left: 0.5rem;
}
.dnb-checkbox--label-position-right .dnb-checkbox__order + .dnb-form-status {
  order: 3;
  vertical-align: top;
  margin-top: 0.5rem;
}
.dnb-checkbox__input:checked ~ .dnb-checkbox__gfx, .dnb-checkbox__input[data-checked=true] ~ .dnb-checkbox__gfx {
  opacity: 1;
  transform: scale(1);
  color: var(--checkbox-color-gfx-on);
}
.dnb-checkbox__input:checked ~ .dnb-checkbox__button, .dnb-checkbox__input[data-checked=true] ~ .dnb-checkbox__button {
  background-color: var(--checkbox-color-background-on);
  border-color: var(--checkbox-color-border-on);
}
.dnb-checkbox__input:not(:checked):not([data-checked=true]) ~ .dnb-checkbox__gfx {
  opacity: 0;
  transform: scale(0.8);
  color: var(--checkbox-color-gfx-off);
}
.dnb-checkbox__input:not(:checked):not([data-checked=true]) ~ .dnb-checkbox__button {
  background-color: var(--checkbox-color-background-off);
  border-color: var(--checkbox-color-border-off);
  transition: background-color 100ms ease-out;
}
.dnb-checkbox__input ~ .dnb-checkbox__indeterminate {
  position: absolute;
  z-index: 5;
  width: var(--checkbox-gfx-width__indeterminate);
  height: var(--checkbox-gfx-height__indeterminate);
  transform: scale(0.85);
  opacity: 0;
  background-color: var(--checkbox-color-gfx-indeterminate);
  border-radius: var(--checkbox-gfx-border-radius__indeterminate);
  transition: opacity 200ms ease-out, transform 200ms ease-out;
}
.dnb-checkbox__input:indeterminate ~ .dnb-checkbox__indeterminate {
  transform: scale(1);
  opacity: 1;
}
.dnb-checkbox__input:indeterminate:checked ~ .dnb-checkbox__button {
  background-color: transparent;
  border-color: var(--checkbox-color-border-off);
}
.dnb-checkbox__input:indeterminate:hover ~ .dnb-checkbox__gfx, .dnb-checkbox__input:indeterminate:checked ~ .dnb-checkbox__gfx, .dnb-checkbox__input:indeterminate:checked:hover ~ .dnb-checkbox__gfx {
  color: transparent;
}
.dnb-checkbox--large .dnb-checkbox__input:indeterminate ~ .dnb-checkbox__indeterminate {
  width: var(--checkbox-gfx-width__indeterminate--large);
  height: var(--checkbox-gfx-height__indeterminate--large);
}
.dnb-checkbox__input[disabled]:checked ~ .dnb-checkbox__button, .dnb-checkbox__input[disabled][data-checked=true] ~ .dnb-checkbox__button {
  background-color: var(--checkbox-color-background-on--disabled);
  border-color: var(--checkbox-color-border-on--disabled);
}
.dnb-checkbox__input[disabled]:checked ~ .dnb-checkbox__gfx, .dnb-checkbox__input[disabled][data-checked=true] ~ .dnb-checkbox__gfx {
  color: var(--checkbox-color-gfx--disabled);
}
.dnb-checkbox__input[disabled]:not(:checked):not([data-checked=true]) ~ .dnb-checkbox__button {
  border-color: var(--checkbox-color-border-off--disabled);
  background-color: var(--checkbox-color-background-off--disabled);
}
.dnb-checkbox__input[disabled]:indeterminate ~ .dnb-checkbox__indeterminate {
  background-color: var(--checkbox-color-gfx__indeterminate--disabled);
}
.dnb-checkbox__input[disabled]:indeterminate ~ .dnb-checkbox__gfx {
  color: transparent;
}
.dnb-checkbox__input:not([disabled]):active ~ .dnb-checkbox__button {
  background-color: var(--checkbox-color-background--active);
  border-color: var(--checkbox-color-border--active);
}
.dnb-checkbox__input:not([disabled]):checked:active ~ .dnb-checkbox__gfx, .dnb-checkbox__input:not([disabled])[data-checked=true]:active ~ .dnb-checkbox__gfx {
  color: var(--checkbox-color-gfx-on);
}
.dnb-checkbox__input:not([disabled]):indeterminate:active ~ .dnb-checkbox__button {
  background-color: var(--checkbox-color-background-indeterminate--active);
  border-color: var(--checkbox-color-border-indeterminate--active);
}
.dnb-checkbox__input:not([disabled]):indeterminate:active ~ .dnb-checkbox__indeterminate {
  background-color: var(--checkbox-color-gfx-indeterminate--active);
}
.dnb-checkbox__input:not([disabled]):not(:active):hover ~ .dnb-checkbox__button {
  border-color: var(--checkbox-color-border-off--hover);
  background-color: var(--checkbox-color-background--hover);
}
.dnb-checkbox__input:hover ~ .dnb-checkbox__gfx {
  color: var(--checkbox-color-gfx--hover);
}
.dnb-checkbox__input:not([disabled]):checked:hover ~ .dnb-checkbox__button, .dnb-checkbox__input:not([disabled])[data-checked=true]:hover ~ .dnb-checkbox__button {
  border: var(--checkbox-border-width--hover) solid var(--checkbox-color-border-on--hover);
}
html[data-whatinput=keyboard] .dnb-checkbox__input:not([disabled]):focus-visible ~ .dnb-checkbox__button {
  border: none;
  background-color: var(--checkbox-color-background--focus);
}
html[data-whatinput=keyboard] .dnb-checkbox__input:not([disabled]):focus-visible ~ .dnb-checkbox__gfx {
  color: var(--checkbox-color-gfx--focus);
}
.dnb-checkbox__input:not([disabled]):focus-visible ~ .dnb-checkbox__button .dnb-checkbox__focus, .dnb-checkbox__input:not([disabled]):active ~ .dnb-checkbox__button .dnb-checkbox__focus {
  display: block;
}
.dnb-checkbox__input:not([disabled]):focus-visible ~ .dnb-checkbox__indeterminate {
  background-color: var(--checkbox-color-gfx--focus);
}
.dnb-checkbox__status--error .dnb-checkbox__input:not([disabled]):not(:active) ~ .dnb-checkbox__button {
  border: none;
}
.dnb-checkbox__status--error .dnb-checkbox__input:not([disabled]):not(:active) ~ .dnb-checkbox__button .dnb-checkbox__focus {
  display: block;
  --border-color: var(--checkbox-color-border--error);
  --border-width: var(--checkbox-border-width);
  box-shadow: 0 0 0 var(--border-width) var(--border-color);
  border-color: transparent;
}
.dnb-checkbox__status--error .dnb-checkbox__input:not([disabled]):hover ~ .dnb-checkbox__button {
  background-color: var(--checkbox-color-background--error--hover);
}
.dnb-checkbox__status--error .dnb-checkbox__input:not([disabled]):hover ~ .dnb-checkbox__button[data-checked=true] {
  border-color: var(--checkbox-color-border--error--hover);
}
.dnb-checkbox__status--error .dnb-checkbox__input:not([disabled]):hover ~ .dnb-checkbox__gfx {
  color: var(--checkbox-color-gfx--error--hover);
}
.dnb-checkbox__status--error .dnb-checkbox__input:not([disabled]):not(:active):not(:hover) ~ .dnb-checkbox__gfx {
  color: var(--checkbox-color-gfx--error);
}
html[data-whatinput=keyboard] .dnb-checkbox__status--error .dnb-checkbox__input:not([disabled]):focus ~ .dnb-checkbox__button, html[data-whatinput=keyboard] .dnb-checkbox__status--error .dnb-checkbox__input:not([disabled]):focus:hover ~ .dnb-checkbox__button {
  border: none;
  background-color: var(--checkbox-color-background--error-contrast);
  --border-color: var(--checkbox-color-border--error);
  --border-width: calc(var(--checkbox-border-width--hover));
  box-shadow: 0 0 0 var(--border-width) var(--border-color);
  border-color: transparent;
}
html[data-whatinput=keyboard] .dnb-checkbox__status--error .dnb-checkbox__input:not([disabled]):focus:hover ~ .dnb-checkbox__gfx {
  color: var(--checkbox-color-gfx--error-contrast);
}
.dnb-checkbox__status--error .dnb-checkbox__input:not([disabled]):not(:active):not(:hover):checked ~ .dnb-checkbox__button, .dnb-checkbox__status--error .dnb-checkbox__input:not([disabled]):not(:active):not(:hover)[data-checked=true] ~ .dnb-checkbox__button {
  background-color: var(--checkbox-color-background-on--error);
}
.dnb-checkbox__status--error .dnb-checkbox__input:indeterminate:not([disabled]):not(:active):not(:hover) ~ .dnb-checkbox__gfx {
  color: var(--checkbox-color-gfx--error-contrast);
  background-color: var(--checkbox-color-gfx--error-contrast);
  border-radius: var(--checkbox-border-radius);
}
.dnb-checkbox__status--error .dnb-checkbox__input:indeterminate:not(:active):not(:hover) ~ .dnb-checkbox__indeterminate {
  background-color: var(--checkbox-color-gfx-indeterminate--error);
}
.dnb-checkbox__status--error .dnb-checkbox__input:not(:indeterminate) ~ .dnb-checkbox__indeterminate {
  background-color: var(--checkbox-color-gfx-indeterminate--error);
}
.dnb-checkbox__status--error .dnb-checkbox__input:indeterminate:hover:not(:active) ~ .dnb-checkbox__indeterminate {
  background-color: var(--checkbox-color-gfx--error--hover);
}
.dnb-checkbox__status--error .dnb-checkbox__input:not([disabled]):not(:active):hover ~ .dnb-checkbox__gfx {
  color: var(--checkbox-color-gfx--error--hover);
}
.dnb-checkbox.dnb-skeleton .dnb-checkbox__input[disabled] ~ .dnb-checkbox__button {
  border-color: var(--skeleton-color);
}
.dnb-checkbox.dnb-skeleton .dnb-checkbox__input[disabled] ~ .dnb-checkbox__button::before {
  border-radius: 0;
}
.dnb-checkbox.dnb-skeleton .dnb-checkbox__input[disabled] ~ .dnb-checkbox__gfx {
  color: var(--skeleton-color);
}
.dnb-checkbox.dnb-skeleton .dnb-checkbox__input[disabled] ~ .dnb-checkbox__indeterminate {
  background-color: var(--skeleton-color);
}"
`;

exports[`Checkbox scss have to match default theme snapshot 1`] = `
"/*
* Checkbox theme
*
*/
.dnb-checkbox {
  --checkbox-color-gfx-on: var(--color-white);
  --checkbox-color-gfx-off: var(--color-white);
  --checkbox-color-background-on: var(--color-sea-green);
  --checkbox-color-background-off: var(--color-white);
  --checkbox-color-border-on: transparent;
  --checkbox-color-border-off: var(--color-sea-green);
  --checkbox-color-gfx--disabled: var(--checkbox-color-gfx-on);
  --checkbox-color-background-on--disabled: var(--color-sea-green-30);
  --checkbox-color-background-off--disabled: var(--color-white);
  --checkbox-color-border-on--disabled: transparent;
  --checkbox-color-border-off--disabled: var(--color-mint-green-50);
  --checkbox-color-gfx__indeterminate--disabled: var(
    --color-mint-green-50
  );
  --checkbox-color-background--active: var(--color-mint-green-50);
  --checkbox-color-border--active: transparent;
  --checkbox-color-gfx--hover: var(--color-sea-green);
  --checkbox-color-background--hover: var(--color-mint-green-50);
  --checkbox-color-border-on--hover: transparent;
  --checkbox-color-border-off--hover: var(--color-sea-green);
  --checkbox-color-gfx--focus: var(--color-emerald-green);
  --checkbox-color-background--focus: var(--color-mint-green-50);
  --checkbox-color-gfx--error: var(--color-fire-red-8);
  --checkbox-color-gfx--error-contrast: var(--color-fire-red);
  --checkbox-color-background-on--error: var(--color-fire-red);
  --checkbox-color-background--error-contrast: var(--color-fire-red-8);
  --checkbox-color-border--error: var(--color-fire-red);
  --checkbox-color-gfx--error--hover: var(--color-fire-red);
  --checkbox-color-background--error--hover: var(--color-fire-red-8);
  --checkbox-color-border--error--hover: var(--color-fire-red);
  --checkbox-color-background-indeterminate: var(--color-white);
  --checkbox-color-gfx-indeterminate: var(--color-sea-green);
  --checkbox-color-background-indeterminate--active: var(--color-white);
  --checkbox-color-border-indeterminate--active: var(
    --color-mint-green-50
  );
  --checkbox-color-gfx-indeterminate--active: var(--color-mint-green-50);
  --checkbox-color-gfx-indeterminate--error: var(--color-fire-red);
  --checkbox-gfx-height__indeterminate: 0.625rem;
  --checkbox-gfx-width__indeterminate: 0.625rem;
  --checkbox-gfx-height__indeterminate--large: 0.875rem;
  --checkbox-gfx-width__indeterminate--large: 0.875rem;
}"
`;
